Revisions visual preview pt2 (visual preview)

khaykov opened the pull request:
This PR adds a visual preview to revision.

I was considering adding a visual preview to `HistoryDetailFragment`, but with the current implementation, the easiest place was `HistoryDetailContainerFragment`, as we can more easily control the state, menu and don't have to access fragments inside `ViewPager` adapter.

PS: I also tried to extract Aztec specific XML attributes to a separate style without much luck :(

To test:

1. Open a revision and using overflow menu in the toolbar switch to visual preview.
2. Make sure visual preview works as expected.

Labels when switching to visual preview:
[![Image from Gyazo](https://i.gyazo.com/8ecdee24f8492256f79f8d483463b568.png)](https://gyazo.com/8ecdee24f8492256f79f8d483463b568)
Label when switching back to HTML preview:
[![Image from Gyazo](https://i.gyazo.com/2ef67dcb5722f1b9ac6459c046c0516a.png)](https://gyazo.com/2ef67dcb5722f1b9ac6459c046c0516a)

Visual preview with a title:
[![Image from Gyazo](https://i.gyazo.com/94a8e7bea4c0775fa0d740cacf3fda79.png)](https://gyazo.com/94a8e7bea4c0775fa0d740cacf3fda79)

Visual preview without a title:
[![Image from Gyazo](https://i.gyazo.com/29c818015306a58cb49f72446b996b93.png)](https://gyazo.com/29c818015306a58cb49f72446b996b93)

iamthomasbishop commented:
This looks pretty solid. Labeling should work fine (HTML preview, Visual preview). The no-title example might be something we address in another pass, but fwiw, I'd match the styling of the initial title placeholder text and the label could be "No Title" in this context.

theck13 commented on WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java:
```diff
@@ -95,47 +194,38 @@ public boolean onOptionsItemSelected(MenuItem item) {
+        } else if (item.getItemId() == R.id.history_toggle_view) {
+            mVisualTitle.setText(isInVisualPreview() ? null : mRevision.getPostTitle());
+            mVisualContent.fromHtml(
+                    isInVisualPreview() ? "" : StringUtils.notNullStr(mRevision.getPostContent()), false);
```
The crossfade from HTML view to visual view looks great!  The crossfade from visual view to HTML view has a minor oddity.  Since we reset the `mVisualTitle` and `mVisualContent` text with `null` and `""` when toggling from visual view to HTML view, those fields become blank showing just a white background at the beginning of the animation creating a blink-like transition.  What about changing the `else if` branch in the `onOptionsItemSelected` method to something like this?

```
        } else if (item.getItemId() == R.id.history_toggle_view) {
            if (isInVisualPreview()) {
                mPreviousButton.setEnabled(mPosition != 0);
                mNextButton.setEnabled(mPosition != mAdapter.getCount() - 1);
            } else {
                mPreviousButton.setEnabled(false);
                mNextButton.setEnabled(false);
                mVisualTitle.setText(mRevision.getPostTitle());
                mVisualContent.fromHtml(StringUtils.notNullStr(mRevision.getPostContent()), false);
            }

            crossfadePreviewViews();
        }
```

That will fix the slightly jarring transition from visual view to HTML view as well the `mPreviousButton` and `mNextButton` state bug.  See the animation below for illustration.

Before|After
-|-
<img src="https://user-images.githubusercontent.com/3827611/48498536-154ddf80-e7fc-11e8-8ef9-3b35a60a60b0.gif" width="320">|<img src="https://user-images.githubusercontent.com/3827611/48498543-1b43c080-e7fc-11e8-85fb-22009c46ebea.gif" width="320">

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java:
```diff
@@ -66,27 +90,102 @@ public static HistoryDetailContainerFragment newInstance(Revision revision, Arra
         }
 
-        mIsFragmentRecreated = savedInstanceState != null;
+        mPosition = mRevisions.indexOf(mRevision);
```
This is an interesting question :) I'm thinking about a situation where `mRevisions` might be `null`, and theoretically, it could happen when something really bad happened. I think there is a benefit in letting app crash (maybe throw exception ourself with the proper error message), so we can notice it earlier and address the issue. What do you think?

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java:
```diff
@@ -66,27 +90,102 @@ public static HistoryDetailContainerFragment newInstance(Revision revision, Arra
+
+        mVisualPreviewContainer = rootView.findViewById(R.id.visual_preview_container);
+
+        boolean isInVisualPreview = mIsFragmentRecreated && savedInstanceState.getBoolean(KEY_IS_IN_VISUAL_PREVIEW);
```
Good point! 👍

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java:
```diff
@@ -66,27 +90,102 @@ public static HistoryDetailContainerFragment newInstance(Revision revision, Arra
+
+        mVisualContent.setImageGetter(new AztecImageLoader(getActivity(), mImageManager, loadingImagePlaceholder));
+        mVisualContent.setKeyListener(null);
+        mVisualContent.setFocusable(false);
```
Good catch! I tried to fix scroll related issue with `setFocusable(false)`, but looks like it's not required anymore.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java:
```diff
@@ -95,47 +194,38 @@ public boolean onOptionsItemSelected(MenuItem item) {
     }
 
-    @Override
-    public void onActivityCreated(@Nullable final Bundle savedInstanceState) {
```
I think it makes more sense to logically group code related to view initialization inside `onCreateView`, and interact with views before they are laid out. We also can avoid `getView() != null` check this way.  I just realized that setting toolbar subtitle depends on activity lifecycle, so I moved that part into `onActivityCreated`.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java:
```diff
@@ -95,47 +194,38 @@ public boolean onOptionsItemSelected(MenuItem item) {
-            refreshHistoryDetail();
-            resetOnPageChangeListener();
-        }
+    private void crossfadePreviewViews() {
```
Thanks! I didn't know about that class.

khaykov commented:
@theck13 Thanks for the throughout review! I think I addressed all of your concerns, except for [this one](https://github.com/wordpress-mobile/WordPress-Android/pull/8593#discussion_r233313719).
As you suggested, I implemented fade in/out of controls and it looks pretty good! @iamthomasbishop, what do you think?
[![Image from Gyazo](https://thumb.gyazo.com/thumb/200/eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpbWciOiJfNjk5ZDg0MWUxYzVhMjY2Njk2NzcyODAzNjMwYTUwMjQifQ.6ScQP48mVv1vtBCXjsGQFeHl52Hf7l5iQl0Dgs8kxso-gif.gif)](https://gyazo.com/b582e80a0865ffadaaa5e27434b61dda)

I also added a No Title placeholder.
[![Image from Gyazo](https://thumb.gyazo.com/thumb/200/eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpbWciOiJfMjNiZjlmMDEzNDMwNjViN2UxZGFkOTFiZjQyMzMyNTgifQ.3C-HFV-uDbUjVldL0B2bCnBqHcyJPijPfuEJwo0Ca3k-png.jpg)](https://gyazo.com/625005b9c4d1e143fc2a80b0489d9c05)

iamthomasbishop commented:
>  I lean toward using "(no title)" as text, but I'll let @iamthomasbishop and @khaykov decide.

This sounds good to me. Clearly we've got a lot of inconsistency which should be cleaned up, but that's beyond the scope of this. Let's roll with what you mentioned, `(no title)`. That will at least match between the HTML/Visual modes w/in this feature (and align w/ core and calypso IIRC).
